framework/src/outputs/CSV.C

```diff
@@ -173,6 +174,15 @@ CSV::getVectorPostprocessorFileName(const std::string & vpp_name,
       file_name << '_' << std::setw(_padding) << std::setprecision(0) << std::setfill('0')
                 << std::right << _linear_iter;
     }
+
+    // These lines enable CSV per-optimization iteration output (Optimization module)
+    if (auto steady = dynamic_cast<Steady *>(_app.getExecutioner()))
+      if (steady->getIterationOutputFlag() && _current_execute_flag == EXEC_TIMESTEP_END)
+      {
+        std::string append = "iteration_" + Moose::stringify(steady->getIterationNumberOutput());
+        file_name << '_' << std::setw(_padding) << std::setprecision(0) << std::setfill('0')
+                  << std::right << append;
+      }
```

Comment:
I think we're crossing over the red line here. It seems like at least some of these APIs can go on the base executioner?

Reply:
We can do that to avoid casting. But I don't know that we have a use case now.

Reply:
It seems like we are realizing @dschwen's fears in https://github.com/idaholab/moose/pull/25011#issuecomment-1650815614. We are somehow in the pretty undesirable middle ground between having kept logic internal to the module vs. having a sufficiently abstract design that it makes sense in the base class